### PR TITLE
Extend readme

### PR DESCRIPTION
**Pull request by brancz**
As we've lately had some questions around this project I wanted to add a little bit of detail on how it works, how we make use of it and adapted the examples to closely resemble this.

@mxinden @squat @s-urbaniak @metalmatze

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,19 +2,22 @@
...
-- If a tenant controls its scrape target configuration the tenant can set
-  arbitrary labels via its relabelling configuration, thereby being able to
-  pollute other tenant's metrics.
+- If a tenant controls its scrape target configuration the tenant can set arbitrary labels via its relabelling configuration, thereby being able to pollute other tenant's metrics.
+
+- If the ingestion configuration [honor_labels](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) is set for a tenant's target, that target can pollute other tenant's metrics as Prometheus respects any labels exposed by the target.
+
+## How does this project work?
+
+What this proxy does is it proxies the `/federate`, `/api/v1/query`, `/api/v1/query_range` prometheus endpoints and ensures that a label particular label is enforced in the particular query.
```
Comment (metalmatze) [inline README.md]:
```suggestion
What this proxy does is it proxies the `/federate`, `/api/v1/query`, `/api/v1/query_range` prometheus endpoints and ensures that a particular label is enforced in the particular query.
```

**Comment by s-urbaniak** [comment]
Thanks a lot @brancz for adding more information here! :+1:

File: README.md [hunk 2]
```diff
@@ -2,19 +2,22 @@
...
+
+In the case of the federate endpoint, it ensures that all selectors passed as matchers to the federate endpoint _must_ contain that exact match of the particular label (and throws away all other matchers for the label). For the two query endpoints, it parses the PromQL expression and modifies all selectors in the same way. The label-key is configured as a flag on the binary and label-value is passed as a query parameter.
+
```
Comment (metalmatze) [inline README.md]: Would be nice to add an more hands-on example, something like:

> Here's an example: We are running our application in the namespace `app` and our application developers should only see metrics for their namespace. This proxy can be configured to enforce that all queries sent to Prometheus always have `{namespace="app"}` as part of their label selectors. Once this label selector is enforced, the returned metrics are only for the developer's `app` namespace.

Comment (brancz) [reply README.md]: Done

File: README.md (hunk 1 above)
Comment (brancz) [reply README.md]: Done

File: README.md [hunk 3]
```diff
@@ -2,19 +2,37 @@
...
+## Example use
 
-- If the ingestion configuration
-  [honor_labels](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config)
-  is set for a tenant's target, that target can pollute other tenant's metrics
-  as Prometheus respects any labels exposed by the target.
+The concrete setup being shipped in OpenShift starting with 4.0: the proxy is configured to work with the label-key: namespace. In order to ensure that this is secure is it paired with the [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) and its URL rewrite functionality, meaning first ServiceAccount token authentication is performed, and then the kube-rbac-proxy authorization to see whether the requesting entity is allowed to retrieve the metrics for the requested namespace. The RBAC role we chose to authorize against is the same as the Kubernetes Resource Metrics API, the reasoning being, if an entity can `kubectl top pod` in a namespace, it can see cAdvisor metrics (container_memory_rss, container_cpu_usage_seconds_total, etc.).
```
Comment (mxinden) [inline README.md]: `s/this is secure is it paired/this is secure it is paired`

**Pull request merged**